File: src/main/java/com/embabel/prepper/agent/Domain.java
package com.embabel.prepper.agent;

import com.embabel.common.ai.prompt.PromptContributor;
import com.fasterxml.jackson.annotation.JsonPropertyDescription;
import jakarta.persistence.*;
import org.jetbrains.annotations.NotNull;
import org.springframework.data.annotation.CreatedDate;
import org.springframework.data.annotation.LastModifiedDate;
import org.springframework.data.jpa.domain.support.AuditingEntityListener;
import org.springframework.lang.NonNull;
import org.springframework.lang.Nullable;

import java.time.LocalDateTime;
import java.util.List;

public abstract class Domain {

    public record Meeting(
            String context,
            String objective,
            @JsonPropertyDescription("A list of participant email addresses or however else we identify them, e.g. 'Roger Daltrey The Who' or 'roger@who.com'")
            List<String> participants
    ) {

        @NonNull
        public String purpose() {
            return "Meeting:\nContext: %s\nObjective: %s\n".formatted(context, objective);
        }
    }

    public record NewContact(
            String name,
            @Nullable String email,
            String writeup
    ) {
    }

    /**
     * Fleshed out participant
     */
    @Entity
    @EntityListeners(AuditingEntityListener.class)
    public static class Contact implements PromptContributor {

        @Id
        @GeneratedValue(strategy = GenerationType.IDENTITY)
        private Long id;

        @Column(nullable = false)
        private String name;

        @Column
        private String email;

        @Column(columnDefinition = "TEXT")
        private String writeup;

        @CreatedDate
        private LocalDateTime createdAt;

        @LastModifiedDate
        private LocalDateTime updatedAt;

        protected Contact() {
        }

        public Contact(NewContact newContact) {
            this.name = newContact.name;
            this.email = newContact.email;
            this.writeup = newContact.writeup;
        }

        @NotNull
        @Override
        public String contribution() {
            return "- %s: %s".formatted(email, writeup);
        }

        public Long getId() {
            return id;
        }

        public void setId(Long id) {
            this.id = id;
        }

        public String getEmail() {
            return email;
        }

        public void setEmail(String email) {
            this.email = email;
        }

        public String getWriteup() {
            return writeup;
        }

        public void setWriteup(String writeup) {
            this.writeup = writeup;
        }

        public String getName() {
            return name;
        }

        public String email() {
            return email;
        }

        public String writeup() {
            return writeup;
        }

        @Override
        public String toString() {
            return "Contact{id=%d, name='%s', email='%s', writeup:\n%s}".formatted(id, name, email, writeup);
        }
    }

    public record Participants(
            List<Contact> participants
    ) implements PromptContributor {

        @NotNull
        @Override
        public String contribution() {
            return participants.stream()
                    .map(Contact::contribution)
                    .collect(java.util.stream.Collectors.joining("\n"));
        }
    }

    public record IndustryAnalysis(
            String analysis
    ) {
    }

    public record MeetingStrategy(
            @JsonPropertyDescription("Complete report with a list of key talking points and strategic questions" +
                    " to ask, to help achieve the meeting's objective")
            String strategy
    ) {
    }

    public record Briefing(
            Meeting meeting,
            Participants participants,
            IndustryAnalysis industryAnalysis,
            MeetingStrategy meetingStrategy,
            String briefing
    ) {
    }

}

File: src/main/kotlin/com/embabel/prepper/agent/Domain.kt
package com.embabel.prepper.agent

import com.embabel.common.ai.prompt.PromptContributor
import com.fasterxml.jackson.annotation.JsonPropertyDescription
import jakarta.persistence.*
import org.springframework.data.annotation.CreatedDate
import org.springframework.data.annotation.LastModifiedDate
import org.springframework.data.jpa.domain.support.AuditingEntityListener
import org.springframework.lang.NonNull
import org.springframework.lang.Nullable
import java.time.LocalDateTime

abstract class Domain {

    data class Meeting(
        val context: String,
        val objective: String,
        @JsonPropertyDescription("A list of participant email addresses or however else we identify them, e.g. 'Roger Daltrey The Who' or 'roger@who.com'")
        val participants: List<String>
    ) {
        @NonNull
        fun purpose(): String = "Meeting:\nContext: $context\nObjective: $objective\n"
    }

    data class NewContact(
        val name: String,
        @field:Nullable val email: String?,
        val writeup: String
    )

    @Entity
    @EntityListeners(AuditingEntityListener::class)
    open class Contact() : PromptContributor {

        @Id
        @GeneratedValue(strategy = GenerationType.IDENTITY)
        open var id: Long? = null

        @Column(nullable = false)
        open lateinit var name: String

        @Column
        open var email: String? = null

        @Column(columnDefinition = "TEXT")
        open var writeup: String? = null

        @CreatedDate
        open var createdAt: LocalDateTime? = null

        @LastModifiedDate
        open var updatedAt: LocalDateTime? = null

        constructor(newContact: NewContact) : this() {
            this.name = newContact.name
            this.email = newContact.email
            this.writeup = newContact.writeup
        }

        override fun contribution(): String = "- $email: $writeup"

        override fun toString(): String =
            "Contact{id=$id, name='$name', email='$email', writeup:\n$writeup}"
    }

    data class Participants(
        val participants: List<Contact>
    ) : PromptContributor {
        override fun contribution(): String = participants.joinToString("\n") { it.contribution() }
    }

    data class IndustryAnalysis(
        val analysis: String
    )

    data class MeetingStrategy(
        @JsonPropertyDescription("Complete report with a list of key talking points and strategic questions to ask, to help achieve the meeting's objective")
        val strategy: String
    )

    data class Briefing(
        val meeting: Meeting,
        val participants: Participants,
        val industryAnalysis: IndustryAnalysis,
        val meetingStrategy: MeetingStrategy,
        val briefing: String
    )
}
